File: 2018/day01_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import Sequence


def solution(content: Sequence[str]) -> int:
    deltas = [int(f) for f in content]
    current = 0
    frequencies = []
    for delta in repeat(deltas):
        if current in frequencies:
            return current
        frequencies.append(current)
        current += delta
    return 0


def repeat(deltas: list[int]):
    index = 0
    while True:
        yield deltas[index % len(deltas)]
        index += 1


def main() -> None:
    with open("day01.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2018/day01-2.ts
function solution(content: string[]) {
  const deltas = content.filter((item) => item).map(Number);
  let current = 0;
  const frequencies: number[] = [];
  const generator = repeat(deltas);
  while (true) {
    const delta = generator.next().value;
    if (frequencies.includes(current)) {
      return current;
    }
    frequencies.push(current);
    current += delta;
  }
}

function* repeat(deltas: number[]): Generator<number, number, unknown> {
  let index = 0;
  while (true) {
    yield deltas[index % deltas.length];
    index += 1;
  }
}

Deno.readTextFile("day01.dat")
  .then((content) => content.split("\n"))
  .then(solution)
  .then(console.log);
